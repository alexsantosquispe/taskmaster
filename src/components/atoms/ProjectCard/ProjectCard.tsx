import { formatToFriendlyDate } from '@/utils/dates.utils';
import cn from 'clsx';
import { Link } from 'react-router-dom';
import { twMerge } from 'tailwind-merge';
import type { ProjectDTO } from '../../../services/apiTypes';
import { ProjectCardContextMenu } from './components/ProjectCardContextMenu/ProjectCardContextMenu';

interface ProjectCardProps extends ProjectDTO {
  isSmall?: boolean;
}

interface CodeBadgeProps {
  color: string;
  code: string;
}

export const CodeBadge = ({ color, code }: CodeBadgeProps) => {
  return (
    <div className="flex items-center gap-1 rounded-md border border-neutral-200 px-1.5 py-1 font-semibold dark:border-white/20">
      <div
        style={{ backgroundColor: color }}
        className={twMerge('h-4 w-4 rounded-full')}
      />
      <span className="text-xs">{code}</span>
    </div>
  );
};

export const ProjectCard = ({
  id,
  name,
  description,
  update_date: lastUpdateDate,
  color,
  code,
  isSmall = false
}: ProjectCardProps) => {
  const timeUpdated = formatToFriendlyDate(lastUpdateDate);

  return (
    <article
      key={id}
      className={twMerge(
        'dark:bg-primary flex h-[8.25rem] flex-col rounded-lg border border-neutral-200 px-4 py-3 text-sm dark:border-white/20',
        cn({ 'h-24': isSmall })
      )}
    >
      <div className="flex flex-1 flex-col gap-3">
        <div className="flex items-center justify-between gap-2">
          <Link
            to={`/projects/${id}`}
            className="hover:text-accent dark:hover:text-accent-dark hover:underline"
          >
            <h3 className="line-clamp-1 text-base font-semibold">{name}</h3>
          </Link>
          <ProjectCardContextMenu />
        </div>

        {!isSmall && (
          <p className="line-clamp-2 flex-1 text-neutral-700 dark:text-neutral-200">
            {description}
          </p>
        )}
      </div>

      <div className="flex items-baseline justify-between gap-2">
        <span className="text-xs text-neutral-500 dark:text-neutral-300">{`updated ${timeUpdated}`}</span>
        <CodeBadge color={color} code={code} />
      </div>
    </article>
  );
};
