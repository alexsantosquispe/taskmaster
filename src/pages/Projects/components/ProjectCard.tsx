export const ProjectCard = () => {
  return (
    <article className="">
      <h3>TaskMaster Mobile App</h3>
      <p>Description</p>
    </article>
  );
};
