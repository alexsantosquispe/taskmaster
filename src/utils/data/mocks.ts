import {
  PRIORITY_TYPES,
  STATUS_TYPES,
  type TaskType
} from '../../models/types';

export const BACKLOG_TASKS: TaskType[] = [
  {
    id: '1',
    title: 'Set up analytics tools',
    description:
      'Install and configure Google Analytics and Hotjar to monitor user behavior and traffic.',
    status: STATUS_TYPES.BACKLOG,
    createdAt: '2025-07-01T10:24:00-04:00',
    lastUpdate: '2025-07-05T14:13:00-04:00',
    priority: PRIORITY_TYPES.MEDIUM,
    progress: 0
  },
  {
    id: '2',
    title: 'Gather user feedback on onboarding',
    description:
      'Collect and analyze user feedback from early access users about the onboarding experience.',
    status: STATUS_TYPES.BACKLOG,
    createdAt: '2025-07-02T11:30:00-04:00',
    lastUpdate: '2025-07-06T09:00:00-04:00',
    priority: PRIORITY_TYPES.HIGH,
    progress: 0
  },
  {
    id: '18',
    title: 'Write project README',
    description:
      'Added complete project documentation including setup, scripts, and folder structure.',
    status: STATUS_TYPES.DONE,
    createdAt: '2025-07-03T13:40:00-04:00',
    lastUpdate: '2025-07-06T10:00:00-04:00',
    priority: PRIORITY_TYPES.LOW,
    progress: 100
  }
];

export const ON_HOLD_TASKS: TaskType[] = [
  {
    id: '7',
    title: 'Finalize pricing model options',
    description:
      'Pricing structure needs stakeholder review before publishing. Waiting for finance team input.',
    status: STATUS_TYPES.ON_HOLD,
    createdAt: '2025-07-03T10:00:00-04:00',
    lastUpdate: '2025-07-09T11:45:00-04:00',
    priority: PRIORITY_TYPES.LOW,
    progress: 75
  }
];

export const IN_PROGRESS_TASKS: TaskType[] = [
  {
    id: '12',
    title: 'Implement authentication flow',
    description:
      'OAuth integration is partially done; currently wiring up token refresh and user session handling.',
    status: STATUS_TYPES.IN_PROGRESS,
    createdAt: '2025-07-06T09:30:00-04:00',
    lastUpdate: '2025-07-10T10:45:00-04:00',
    priority: PRIORITY_TYPES.HIGH,
    progress: 45
  },
  {
    id: '13',
    title: 'Integrate Supabase for data storage',
    description:
      'Connection established. Working on CRUD operations and testing permissions for different roles.',
    status: STATUS_TYPES.IN_PROGRESS,
    createdAt: '2025-07-08T14:00:00-04:00',
    lastUpdate: '2025-07-10T09:30:00-04:00',
    priority: PRIORITY_TYPES.HIGH,
    progress: 80
  }
];

export const REVIEW_TASKS: TaskType[] = [
  {
    id: '8',
    title: 'Update brand guidelines',
    description:
      'Rebranding paused until feedback from the design lead is received.',
    status: STATUS_TYPES.REVIEW,
    createdAt: '2025-07-02T14:20:00-04:00',
    lastUpdate: '2025-07-06T17:00:00-04:00',
    priority: PRIORITY_TYPES.MEDIUM,
    progress: 100
  },
  {
    id: '9',
    title: 'Legal review for terms of service',
    description:
      'Sent draft to legal team. Awaiting confirmation before moving forward.',
    status: STATUS_TYPES.REVIEW,
    createdAt: '2025-07-05T13:10:00-04:00',
    lastUpdate: '2025-07-08T12:30:00-04:00',
    priority: PRIORITY_TYPES.HIGH,
    progress: 100
  },
  {
    id: '19',
    title: 'Create logo and favicon',
    description:
      'Exported multiple sizes and formats for the logo. Favicon added to HTML head.',
    status: STATUS_TYPES.DONE,
    createdAt: '2025-07-04T08:00:00-04:00',
    lastUpdate: '2025-07-07T09:20:00-04:00',
    priority: PRIORITY_TYPES.MEDIUM,
    progress: 100
  }
];

export const DONE_TASKS: TaskType[] = [
  {
    id: '15',
    title: 'Deploy staging environment',
    description:
      'Staging environment successfully deployed and tested with mock data.',
    status: STATUS_TYPES.DONE,
    createdAt: '2025-06-30T11:00:00-04:00',
    lastUpdate: '2025-07-03T15:30:00-04:00',
    priority: PRIORITY_TYPES.HIGH,
    progress: 100
  },
  {
    id: '16',
    title: 'Fix navbar responsiveness',
    description:
      'Navbar layout now works correctly across all screen sizes and breakpoints.',
    status: STATUS_TYPES.DONE,
    createdAt: '2025-07-01T10:00:00-04:00',
    lastUpdate: '2025-07-04T12:15:00-04:00',
    priority: PRIORITY_TYPES.MEDIUM,
    progress: 100
  },
  {
    id: '17',
    title: 'Set up CI/CD pipeline',
    description:
      'GitHub Actions configured to run tests and deploy to Vercel on push to main.',
    status: STATUS_TYPES.DONE,
    createdAt: '2025-07-02T09:30:00-04:00',
    lastUpdate: '2025-07-05T14:45:00-04:00',
    priority: PRIORITY_TYPES.HIGH,
    progress: 100
  },

  {
    id: '20',
    title: 'Test form validation logic',
    description:
      'Test cases written and validated for all form fields using React Hook Form.',
    status: STATUS_TYPES.DONE,
    createdAt: '2025-07-05T10:45:00-04:00',
    lastUpdate: '2025-07-08T11:30:00-04:00',
    priority: PRIORITY_TYPES.HIGH,
    progress: 100
  }
];
