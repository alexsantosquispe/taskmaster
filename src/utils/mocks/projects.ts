import type { ProjectDTO } from '../../services/apiTypes';

export const PROJECTS: ProjectDTO[] = [
  {
    id: '4e393d82-2ec9-46e6-89a1-f42aca499e24',
    name: 'Spotify Clone App',
    description: 'This is new project with react native',
    create_date: '2025-08-22T17:24:45.746846+00:00',
    color: '#2563eb',
    code: 'SCA',
    update_date: '2025-08-23T20:00:00.746846+00:00'
  },
  {
    id: 'ac2eed41-680b-4cd9-b079-7db1e4fac09e',
    name: 'Resume Builder V1',
    description:
      'Tihs project is a web app to create an resume and check the how it can be improved with AI',
    create_date: '2025-08-21T03:59:50.116667+00:00',
    color: '#f97316',
    code: 'RESB',
    update_date: '2025-08-21T03:59:50.116667+00:00'
  },
  {
    id: 'd9587eaa-1000-49db-8ce9-183fa47a1696',
    name: 'Task Master Mobile App',
    description: 'This is a mobile version with react native',
    create_date: '2025-08-21T03:40:21.137062+00:00',
    color: '#eab308',
    code: 'TMMA',
    update_date: '2025-08-21T03:40:21.137062+00:00'
  },
  {
    id: '5082b2a0-1e2a-43ca-9661-a12260520a15',
    name: 'Web Track App',
    description: 'Agile project management tool',
    create_date: '2025-08-18T23:51:48.391714+00:00',
    color: '#8B5CF6',
    code: 'DEV',
    update_date: '2025-08-18T23:51:48.391714+00:00'
  }
];
