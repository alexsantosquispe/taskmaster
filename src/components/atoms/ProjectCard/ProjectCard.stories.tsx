import type { Meta, StoryObj } from '@storybook/react';

import { PROJECTS } from '@/utils/mocks/projects';
import { ProjectCard } from './ProjectCard';

const meta = {
  title: 'Atoms/ProjectCard',
  component: ProjectCard,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof ProjectCard>;

export default meta;

type Story = StoryObj<typeof ProjectCard>;

export const Default: Story = () => {
  const proj1 = { ...PROJECTS[0] };
  const proj2 = { ...PROJECTS[1] };

  return (
    <div className="grid grid-cols-2 gap-16">
      <div className="flex flex-col gap-4">
        <span>Regular size</span>
        <ProjectCard {...proj1} />
      </div>

      <div className="flex flex-col gap-4">
        <span>Small size</span>
        <ProjectCard {...proj2} isSmall={true} />
      </div>
    </div>
  );
};

Default.args = {};
