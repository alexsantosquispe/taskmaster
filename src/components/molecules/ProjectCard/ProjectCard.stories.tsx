import type { Meta, StoryObj } from '@storybook/react';

import { PROJECTS } from '@/utils/mocks/projects';
import { TestWrapper } from '@/utils/wrappers.utils';
import { ProjectCard } from './ProjectCard';

const meta = {
  title: 'Molecules/ProjectCard',
  component: ProjectCard,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof ProjectCard>;

export default meta;

type Story = StoryObj<typeof ProjectCard>;

export const Default: Story = () => {
  const proj1 = { ...PROJECTS[0] };
  const proj2 = { ...PROJECTS[1] };

  return (
    <div className="grid grid-cols-2 gap-16">
      <div className="flex flex-col gap-4">
        <span>Regular size</span>
        <TestWrapper>
          <ProjectCard {...proj1} />
        </TestWrapper>
      </div>

      <div className="flex flex-col gap-4">
        <span>Small size</span>
        <TestWrapper>
          <ProjectCard {...proj2} isSmall={true} />
        </TestWrapper>
      </div>
    </div>
  );
};

Default.args = {};
