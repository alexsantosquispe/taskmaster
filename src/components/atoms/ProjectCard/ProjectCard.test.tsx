import { TestWrapper } from '@/utils/testing/unitTest.util';
import { render } from '@testing-library/react';
import { PROJECTS } from '../../../utils/mocks/projects';
import { ProjectCard } from './ProjectCard';

describe('ProjectCard', () => {
  const props = { ...PROJECTS[0] };

  beforeEach(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date(2025, 7, 24));
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should render the component correctly', () => {
    let component = render(
      <TestWrapper>
        <ProjectCard {...props} />
      </TestWrapper>
    );

    expect(component).toMatchSnapshot();

    component = render(
      <TestWrapper>
        <ProjectCard {...props} isSmall={true} />
      </TestWrapper>
    );

    expect(component).toMatchSnapshot();
  });
});
