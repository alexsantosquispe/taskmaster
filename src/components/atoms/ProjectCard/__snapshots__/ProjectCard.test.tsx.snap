// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProjectCard should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="dark:bg-primary flex h-[8.25rem] flex-col rounded-lg border border-neutral-200 px-4 py-3 text-sm dark:border-white/20"
      >
        <div
          class="flex flex-1 flex-col gap-3"
        >
          <a
            class="hover:text-accent dark:hover:text-accent-dark hover:underline"
            href="/projects/4e393d82-2ec9-46e6-89a1-f42aca499e24"
          >
            <h3
              class="line-clamp-1 text-base font-semibold"
            >
              Spotify Clone App
            </h3>
          </a>
          <p
            class="line-clamp-2 flex-1 text-neutral-700 dark:text-neutral-200"
          >
            This is new project with react native
          </p>
        </div>
        <div
          class="flex items-baseline justify-between gap-2"
        >
          <span
            class="text-xs text-neutral-500 dark:text-neutral-300"
          >
            updated 4 hours ago
          </span>
          <div
            class="flex items-center gap-1 rounded-md border border-neutral-200 px-1.5 py-1 font-semibold dark:border-white/20"
          >
            <div
              class="h-4 w-4 rounded-full"
              style="background-color: rgb(37, 99, 235);"
            />
            <span
              class="text-xs"
            >
              SCA
            </span>
          </div>
        </div>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="dark:bg-primary flex h-[8.25rem] flex-col rounded-lg border border-neutral-200 px-4 py-3 text-sm dark:border-white/20"
    >
      <div
        class="flex flex-1 flex-col gap-3"
      >
        <a
          class="hover:text-accent dark:hover:text-accent-dark hover:underline"
          href="/projects/4e393d82-2ec9-46e6-89a1-f42aca499e24"
        >
          <h3
            class="line-clamp-1 text-base font-semibold"
          >
            Spotify Clone App
          </h3>
        </a>
        <p
          class="line-clamp-2 flex-1 text-neutral-700 dark:text-neutral-200"
        >
          This is new project with react native
        </p>
      </div>
      <div
        class="flex items-baseline justify-between gap-2"
      >
        <span
          class="text-xs text-neutral-500 dark:text-neutral-300"
        >
          updated 4 hours ago
        </span>
        <div
          class="flex items-center gap-1 rounded-md border border-neutral-200 px-1.5 py-1 font-semibold dark:border-white/20"
        >
          <div
            class="h-4 w-4 rounded-full"
            style="background-color: rgb(37, 99, 235);"
          />
          <span
            class="text-xs"
          >
            SCA
          </span>
        </div>
      </div>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ProjectCard should render the component correctly 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="dark:bg-primary flex h-[8.25rem] flex-col rounded-lg border border-neutral-200 px-4 py-3 text-sm dark:border-white/20"
      >
        <div
          class="flex flex-1 flex-col gap-3"
        >
          <a
            class="hover:text-accent dark:hover:text-accent-dark hover:underline"
            href="/projects/4e393d82-2ec9-46e6-89a1-f42aca499e24"
          >
            <h3
              class="line-clamp-1 text-base font-semibold"
            >
              Spotify Clone App
            </h3>
          </a>
          <p
            class="line-clamp-2 flex-1 text-neutral-700 dark:text-neutral-200"
          >
            This is new project with react native
          </p>
        </div>
        <div
          class="flex items-baseline justify-between gap-2"
        >
          <span
            class="text-xs text-neutral-500 dark:text-neutral-300"
          >
            updated 4 hours ago
          </span>
          <div
            class="flex items-center gap-1 rounded-md border border-neutral-200 px-1.5 py-1 font-semibold dark:border-white/20"
          >
            <div
              class="h-4 w-4 rounded-full"
              style="background-color: rgb(37, 99, 235);"
            />
            <span
              class="text-xs"
            >
              SCA
            </span>
          </div>
        </div>
      </article>
    </div>
    <div>
      <article
        class="dark:bg-primary flex flex-col rounded-lg border border-neutral-200 px-4 py-3 text-sm dark:border-white/20 h-24"
      >
        <div
          class="flex flex-1 flex-col gap-3"
        >
          <a
            class="hover:text-accent dark:hover:text-accent-dark hover:underline"
            href="/projects/4e393d82-2ec9-46e6-89a1-f42aca499e24"
          >
            <h3
              class="line-clamp-1 text-base font-semibold"
            >
              Spotify Clone App
            </h3>
          </a>
        </div>
        <div
          class="flex items-baseline justify-between gap-2"
        >
          <span
            class="text-xs text-neutral-500 dark:text-neutral-300"
          >
            updated 4 hours ago
          </span>
          <div
            class="flex items-center gap-1 rounded-md border border-neutral-200 px-1.5 py-1 font-semibold dark:border-white/20"
          >
            <div
              class="h-4 w-4 rounded-full"
              style="background-color: rgb(37, 99, 235);"
            />
            <span
              class="text-xs"
            >
              SCA
            </span>
          </div>
        </div>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="dark:bg-primary flex flex-col rounded-lg border border-neutral-200 px-4 py-3 text-sm dark:border-white/20 h-24"
    >
      <div
        class="flex flex-1 flex-col gap-3"
      >
        <a
          class="hover:text-accent dark:hover:text-accent-dark hover:underline"
          href="/projects/4e393d82-2ec9-46e6-89a1-f42aca499e24"
        >
          <h3
            class="line-clamp-1 text-base font-semibold"
          >
            Spotify Clone App
          </h3>
        </a>
      </div>
      <div
        class="flex items-baseline justify-between gap-2"
      >
        <span
          class="text-xs text-neutral-500 dark:text-neutral-300"
        >
          updated 4 hours ago
        </span>
        <div
          class="flex items-center gap-1 rounded-md border border-neutral-200 px-1.5 py-1 font-semibold dark:border-white/20"
        >
          <div
            class="h-4 w-4 rounded-full"
            style="background-color: rgb(37, 99, 235);"
          />
          <span
            class="text-xs"
          >
            SCA
          </span>
        </div>
      </div>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
